Libraries/LibIPC/TransportSocket.cpp

```diff
@@ -153,35 +159,77 @@ struct MessageHeader {
 };
 
 void TransportSocket::post_message(Vector<u8> const& bytes_to_write, Vector<NonnullRefPtr<AutoCloseFileDescriptor>> const& fds)
 {
-    Vector<u8> message_buffer;
-    message_buffer.resize(sizeof(MessageHeader) + bytes_to_write.size());
-    MessageHeader header;
-    header.payload_size = bytes_to_write.size();
-    header.fd_count = fds.size();
-    header.type = MessageHeader::Type::Payload;
-    memcpy(message_buffer.data(), &header, sizeof(MessageHeader));
-    memcpy(message_buffer.data() + sizeof(MessageHeader), bytes_to_write.data(), bytes_to_write.size());
+    auto total_fds = fds.size();
+    VERIFY(total_fds <= NumericLimits<u32>::max());
 
     for (auto const& fd : fds)
         m_fds_retained_until_received_by_peer.enqueue(fd);
 
     auto raw_fds = Vector<int, 1> {};
-    auto num_fds_to_transfer = fds.size();
-    if (num_fds_to_transfer > 0) {
-        raw_fds.ensure_capacity(num_fds_to_transfer);
-        for (auto& owned_fd : fds) {
+    if (total_fds > 0) {
+        raw_fds.ensure_capacity(total_fds);
+        for (auto const& owned_fd : fds) {
             raw_fds.unchecked_append(owned_fd->value());
         }
     }
 
+    if (total_fds < Core::LocalSocket::MAX_TRANSFER_FDS) {
+        auto message_buffer = MessageHeader::encode_with_payload({
+                                                                     .type = MessageHeader::Type::Payload,
+                                                                     .payload_size = static_cast<u32>(bytes_to_write.size()),
+                                                                     .fd_count = static_cast<u32>(total_fds),
+                                                                 },
+            bytes_to_write);
+        m_send_queue->enqueue_message(move(message_buffer), move(raw_fds));
+        return;
+    }
+
+    // Send MAX_TRANSFER_FDS in FDs only messages until we are below the line.
+    // Send the payload afterward, to reduce churn waiting for extra fds while we have a payload ready to decode
+
+    auto fds_left = total_fds;
+    do {
+        auto message_buffer = MessageHeader::encode_with_payload({
+                                                                     .type = MessageHeader::Type::FileDescriptorsOnly,
+                                                                     .payload_size = 0,
+                                                                     .fd_count = static_cast<u32>(Core::LocalSocket::MAX_TRANSFER_FDS),
+                                                                 },
+            {});
+
+        Vector<int> fds_to_send { raw_fds.span().slice(0, Core::LocalSocket::MAX_TRANSFER_FDS) };
+        raw_fds.remove(0, Core::LocalSocket::MAX_TRANSFER_FDS);
+        m_send_queue->enqueue_message(move(message_buffer), move(fds_to_send));
+
+        fds_left -= Core::LocalSocket::MAX_TRANSFER_FDS;
+
+        ASSERT(raw_fds.size() == fds_left);
+    } while (fds_left > Core::LocalSocket::MAX_TRANSFER_FDS);
```

Comment:
I'm pretty sure 100% of this extra calls to enqueue_message is a waste of time, because of the SOCK_STREAM nature of the underlying socket :(